components/playlist/browser/playlist_tab_helper.cc

```diff
@@ -154,12 +156,48 @@ std::u16string PlaylistTabHelper::GetSavedFolderName() {
   return base::UTF8ToUTF16(service_->GetPlaylist(parent_id)->name);
 }
 
+bool PlaylistTabHelper::ShouldExtractMediaFromBackgroundWebContents() const {
+  return service_->ShouldExtractMediaFromBackgroundWebContents(found_items());
+}
+
+bool PlaylistTabHelper::IsExtractingMediaFromBackgroundWebContents() const {
+  return media_extracted_from_background_web_contents_callbacks_.size();
+}
+
+void PlaylistTabHelper::ExtractMediaFromBackgroundWebContents(
+    base::OnceCallback<void(bool)> extracted_callback) {
+  if (!ShouldExtractMediaFromBackgroundWebContents()) {
+    std::move(extracted_callback).Run(true);
+    return;
+  }
+
+  media_extracted_from_background_web_contents_callbacks_.push_back(
+      std::move(extracted_callback));
+  media_extraction_from_background_web_contents_timer_.Start(
+      FROM_HERE, base::Seconds(10),
+      base::BindOnce(
+          &PlaylistTabHelper::OnMediaExtractionFromBackgroundWebContentsTimeout,
+          weak_ptr_factory_.GetWeakPtr()));
+  ExtractMediaFromBackgroundContents();
+}
+
+void PlaylistTabHelper::RequestAsyncExecuteScript(
+    int32_t world_id,
+    const std::u16string& script,
+    base::OnceCallback<void(base::Value)> cb) {
+  GetRemote(web_contents()->GetPrimaryMainFrame())
+      ->RequestAsyncExecuteScript(
+          world_id, script, blink::mojom::UserActivationOption::kActivate,
+          blink::mojom::PromiseResultOption::kAwait, std::move(cb));
+}
+
 void PlaylistTabHelper::DidFinishNavigation(
     content::NavigationHandle* navigation_handle) {
   DVLOG(2) << __FUNCTION__;
 
-  if (auto old_url = std::exchange(target_url, web_contents()->GetVisibleURL());
-      old_url == target_url) {
+  if (auto old_url =
+          std::exchange(target_url_, web_contents()->GetVisibleURL());
```

Comment:
You should be using `PrimaryPageChanged` here. The current code is also missing `IsInPrimaryMainFrame`, `IsSameDocument`, etc... which are not necessary in `PrimaryPageChanged`
https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/web_contents_observer.h;l=167?q=web_contents_observer.h&ss=chromium

Reply:
Thanks, filed a new issue for it https://github.com/brave/brave-browser/issues/36005

Reply:
Just a heads-up that `WebContentsObserver::PrimaryPageChanged()` breaks `PlaylistTabHelper`, as it doesn't get called on same-document navigations (on which we have to reset found items in `PlaylistTabHelper`). Including a fix in my current refactor works.

Reply:
Yeah, it's broken in Youtube.